dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/RelationAnalysis.java

```diff
@@ -976,7 +976,8 @@ public Knowledge visitSameScope(Relation rel, String specificScope) {
             List<Event> events = new ArrayList<>();
             events.addAll(program.getThreadEvents(Load.class));
             events.addAll(program.getThreadEvents(Store.class));
-            events.addAll(program.getThreadEvents(Fence.class));
+            events.addAll(program.getThreadEvents(GenericVisibleEvent.class));
+            events.removeIf(e -> e instanceof GenericVisibleEvent && !e.hasTag(Tag.FENCE));
```

Comment:
Here I would also go over all events and take those that have `Visible` and `Fence` tag.